cmd/snap/cmd_model.go

```diff
@@ -0,0 +1,228 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+)
+
+var (
+	shortModelHelp = i18n.G("Get the active model for this device")
+	longModelHelp  = i18n.G(`
+The model command returns the active model assertion information for this 
+device.	
+
+By default, only the primary keys of the headers are included in the output, but
+this can be expanded to include all of an assertion's headers non-meta headers.
+
+Similarly, the default model assertion can alternatively be replaced with the 
+serial assertion.
+`)
+
+	// this list is a "nice" "human" "readable" "ordering" of headers to print
+	// off, sorted in lexographical order with meta headers and primary key
+	// headers removed, and big nasty keys such as device-key-sha3-384 and
+	// device-key at the bottom
+	// it also contains both serial and model assertion headers, but we
+	// follow the same code path for both assertion types and some of the
+	// headers are shared between the two, so it still works out correctly
+	niceOrdering = []string{
+		"architecture",
+		"base",
+		"classic",
+		"display-name",
+		"gadget",
+		"kernel",
+		"revision",
+		"timestamp",
+		"required-snaps",
+		"device-key-sha3-384",
+		"device-key",
+	}
+)
+
+type cmdModel struct {
+	waitMixin
+	timeMixin
+	Serial    bool `long:"serial"`
+	Verbose   bool `long:"verbose"`
+	Assertion bool `long:"assertion"`
+}
+
+func init() {
+	addCommand("model",
+		shortModelHelp,
+		longModelHelp,
+		func() flags.Commander {
+			return &cmdModel{}
+		}, timeDescs.also(waitDescs.also(map[string]string{
+			"assertion": i18n.G("Print the raw assertion."),
+			"verbose":   i18n.G("Print all specific assertion fields."),
+			"serial": i18n.G(
+				"Print the serial assertion instead of the model assertion."),
+		})),
+		[]argDesc{},
+	)
+}
+
+func (x *cmdModel) Execute(args []string) error {
+	var assertion asserts.Assertion
+	var err error
+	if x.Serial {
+		assertion, err = x.client.CurrentSerialAssertion()
+	} else {
+		assertion, err = x.client.CurrentModelAssertion()
+	}
+
+	if err != nil && client.IsAssertionNotFoundError(err) {
+		// device is not registered yet - print helpful message to stdout and
+		// return err to set exit code non-zero
+		fmt.Fprintln(Stdout, "device not ready - no assertion found")
+		return err
+	} else if err != nil {
+		return err
```

Comment:
I think this if block would be easier to read as
```go
if err != nil {
        if client.IsAssertionNotFoundError(err) {
                fmt.Fprintln(...)
        }
        return err
}
```

Reply:
yes that is simpler, thanks for the suggestion

Reply:
... not a fan of the switch you used instead of this btw.

You could use the exact same pattern we'd agreed,
```go
if err != nil {
        if client.IsAssertionNotFoundError(err) {
                return fmt.Errorf(...)
        }
        return err
}
```
or even
```go
if err != nil {
        if client.IsAssertionNotFoundError(err) {
                err = fmt.Errorf(...)
        }
        return err
}
```
without it being as jarring as the `switch` is right now